src/aws-cpp-sdk-core/include/aws/core/utils/ResourceManager.h

```diff
@@ -74,6 +74,31 @@ namespace Aws
                 return resource;
             }
 
+            /**
+             * Returns a resource with exclusive ownership or a nullptr.
+             * If resource is available within the wait timeout then the resource is returned.
+             * otherwise (if the timeout has expired or container is shutdown) the nullptr is returned.
+             * You must call Release on the resource when you are finished.
+             * This method is enabled only for pointer RESOURCE_TYPE type.
+             *
+             * @return instance of RESOURCE_TYPE, nullptr if the resource manager is being shutdown
+             */
+            template<typename std::enable_if<std::is_pointer<RESOURCE_TYPE>::value>::type * = nullptr>
+            RESOURCE_TYPE TryAcquire(const uint64_t timeoutMs) {
+              std::unique_lock<std::mutex> locker(m_queueLock);
+              const bool isNotEmpty = m_semaphore.wait_for(locker, std::chrono::milliseconds(timeoutMs),
+                                                           [&]() { return m_shutdown.load() || !m_resources.empty(); });
+
+              if (m_shutdown || !isNotEmpty) {
```

Comment:
nit `!` and `isNot` is kind of confusing could we just use `isEmpty` instead?

Reply:
wait_for returns an enum. 
https://en.cppreference.com/w/cpp/thread/condition_variable/wait_for
Better to use this .
enum class cv_status {
    no_timeout,  => 0
    timeout    => 1
};

doesm't that mean return on !isNotEmpty is actually returning on no timeout. Is that the intention?

Reply:
Also, isNotEmpty can be renamed to isTimeOut for clarity

Reply:
Can we add a test case to test the timeout scenario, which should be easy to replicate

Reply:
No it doesn't return an enum. Check the spec you linked for the (2) with predicate.

Reply:
that's right. looks ok then.  if you can add a test case that will be great

Reply:
It is a really hard to test this piece of code: curl wrapper is hard-coded to a real c curl APIs, handle container is private and cannot be replaced.
I have a 1y old draft of curl wrapper refactoring: https://github.com/aws/aws-sdk-cpp/tree/sr/curlMulti2 which I still want to have merged one day. I can add curl wrapper unit tests there.

Reply:
cool, thanks

Reply:
Added a test on ResourceManager template class.